python-lib/osisoft_client.py

```diff
@@ -43,6 +43,144 @@ def get_auth(self, auth_type, username, password):
             return None
 
+    def get_all_rows_from_webid(self, webid, data_type, start_date=None, end_date=None,
+                                interval=None, sync_time=None, boundary_type=None, selected_fields=None,
+                                can_raise=True, endpoint_type="event_frames", search_full_hierarchy=None, max_count=None):
+        # Split the time range until no more HTTP 400
+        not_happy = True
+        previous_item_timestamp = False
+        while not_happy:
+            logger.info("Attempting download from {} to {}".format(start_date, end_date))
+            ret = self.get_row_from_webid(webid, data_type, start_date=start_date, end_date=end_date,
+                                          interval=interval, sync_time=sync_time, boundary_type=boundary_type, selected_fields=selected_fields,
+                                          can_raise=can_raise, endpoint_type=endpoint_type, search_full_hierarchy=search_full_hierarchy, max_count=max_count)
+            """
+            ret = self.get_row_from_item(item, data_type, start_date=start_date, end_date=end_date, interval=interval,
+                          sync_time=sync_time, boundary_type=boundary_type, can_raise=can_raise, object_id=None,
+                          search_full_hierarchy=search_full_hierarchy, max_count=max_count):
+            """
+            counter = 0
+            try:
+                first_item = next(ret)
+                if not previous_item_timestamp or previous_item_timestamp != first_item.get("Timestamp"):
+                    logger.info("Yielding first item")
+                    yield first_item
+                counter += 1
+                for first_item in ret:
+                    yield first_item
+                    counter += 1
+            except Exception as err:
+                if "Error 400" in "{}".format(err):
+                    logger.warning("The time range {} -> {} is too large, splitting the job in two".format(start_date, end_date))
+                    start_timestamp = self.parse_pi_time(start_date)
+                    end_timestamp = self.parse_pi_time(end_date)
+                    new_time_range = (end_timestamp - start_timestamp) / 2
+                    half_time_iso = epoch_to_iso(start_timestamp + new_time_range)
+                    first_half_items = self.get_all_rows_from_webid(
+                        webid, data_type, start_date=epoch_to_iso(start_timestamp), end_date=half_time_iso,
+                        interval=interval, sync_time=sync_time, boundary_type=boundary_type, selected_fields=selected_fields,
+                        can_raise=can_raise, endpoint_type=endpoint_type, search_full_hierarchy=search_full_hierarchy, max_count=max_count
+                    )
+                    for item in first_half_items:
+                        yield item
+                    logger.info("Successfuly retrieved first half of {} to {}".format(epoch_to_iso(start_timestamp), half_time_iso))
```

Comment:
Minor: Typo in log line (several places):
Successfuly -> Successfully

Reply:
👍 fixed in [0aaa2da](https://github.com/dataiku/dss-plugin-osisoft/pull/43/commits/0aaa2da0f1111b45e59a731ff589dcfcb5c9b3b3)